### PR TITLE
DAOS-15089 object: refine sgl pre-process

### PR DESCRIPTION
**Pull request by liuxuezhao**
Posix might provide sgl with iov of (base NULL, len 0) for readv/writev, refine obj_sgls_dup to process that case.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by ashleypittman** [comment]
Can you remove the https://github.com/daos-stack/daos/blob/master/src/client/dfuse/il/int_write.c#L96-L98 and matching code from read with this PR?

File: ci/docker_nlt.sh [hunk 1]
```diff
@@ -10,6 +10,7 @@ set -x
 
 . utils/sl/setup_local.sh
 
+ps auwx
```
Comment (wangdi1) [inline ci/docker_nlt.sh:13]: for debugging purpose?

File: src/object/cli_obj.c [hunk 2]
```diff
@@ -4535,14 +4535,14 @@ obj_size_fetch_cb(const struct dc_object *obj, struct obj_auxi_args *obj_auxi)
  * For that case, we duplicate the sgls and make its iov_buf_len = iov_len.
  */
 static int
-obj_update_sgls_dup(struct obj_auxi_args *obj_auxi, daos_obj_update_t *args)
+obj_sgls_dup(struct obj_auxi_args *obj_auxi, daos_obj_update_t *args, bool update)
```
Comment (wangdi1) [inline src/object/cli_obj.c:4538]: Just curious, why do we need this duplication.  daos client stack can not handle iov_buf == NULL/iov_lent = 0?

**Review by wangdi1 — APPROVED** [review]

File: src/object/cli_obj.c (hunk 2 above)
Comment (liuxuezhao) [reply src/object/cli_obj.c:4538]: will hit some failures in mercury if with that kind of iov

File: ci/docker_nlt.sh (hunk 1 above)
Comment (liuxuezhao) [reply ci/docker_nlt.sh:13]: right to debug a NLT failure

File: src/object/cli_obj.c (hunk 2 above)
Comment (wangdi1) [reply src/object/cli_obj.c:4538]: Just curious can we fix this to allow some iov_buf in sgls to be iov_buf == NULL/iov_len = 0  in mercury? or I miss sth. @soumagne

Comment (liuxuezhao) [reply src/object/cli_obj.c:4538]: > Just curious can we fix this to allow some iov_buf in sgls to be iov_buf == NULL/iov_len = 0 in mercury? or I miss sth. @soumagne

To be clear, I am not very sure if the failure is caused by mercury.
The failure reported in daos log is -
ERROR: daos_engine:1 02/05-01:57:03.55 boro-41 DAOS[3615093/3/112] object ERR  src/object/srv_obj.c:449 bulk_transfer_sgl() Remote bulk isn't large enough. local_sz:43692, remote_sz:98304, remote_off:87380, DER_OVERFLOW(-1017): 'Data too long for defined data type or buffer size'
ERROR: daos_engine:1 02/05-01:57:03.55 boro-41 DAOS[3615093/3/112] object ERR  src/object/srv_obj.c:578 obj_bulk_transfer() bulk_transfer_sgl i 0, skip_nr 0 failed, DER_OVERFLOW(-1017): 'Data too long for defined data type or buffer size'
...
and 
ERROR: daos_engine:1 02/05-01:48:02.12 boro-35 DAOS[4138276/3/16267] bio  ERR  src/bio/bio_buffer.c:475 copy_one() Invalid iov[0] 0/0 1
ERROR: daos_engine:1 02/05-01:48:02.12 boro-35 DAOS[4138276/3/16267] object ERR  src/object/srv_obj.c:1653 obj_local_rw_internal() 2251804108652548.1.1.1 data transfer failed, dma 0: DER_INVAL(-1003): 'Invalid parameters'
02/05-01:48:02.12 boro-35 DAOS[4138276/3/16267] object ERR  src/object/srv_obj.c:152 obj_rw_complete() 2251804108652548.1.1.1 Fetch end failed: DER_INVAL(-1003): 'Invalid parameters'
.

This patch just dup+removes the (NULL, 0) iov that is the simplest fix, the dup is not needed for common case in critical path.

Comment (NiuYawei) [reply src/object/cli_obj.c:4538]: what test caused such errors? I don't quite see how the errors are related to this PR.

Comment (liuxuezhao) [reply src/object/cli_obj.c:4538]: > what test caused such errors? I don't quite see how the errors are related to this PR.

the changed test case in this PR ("daos_test -i -u 12") can reproduce the failure.

**Review by NiuYawei — APPROVED** [review]

File: ci/docker_nlt.sh (hunk 1 above)
Comment (ashleypittman) [reply ci/docker_nlt.sh:13]: Now fixed in https://github.com/daos-stack/daos/pull/13726

**Comment by liuxuezhao** [comment]
> Can you remove the https://github.com/daos-stack/daos/blob/master/src/client/dfuse/il/int_write.c#L96-L98 and matching code from read with this PR?

as we discussed offline, i think that two lines of code is useful that can save the sgl dup in daos obj layer in this PR. This pr can handle other case the passed in 0 iov although.

**Comment by ashleypittman** [comment]
> > Can you remove the https://github.com/daos-stack/daos/blob/master/src/client/dfuse/il/int_write.c#L96-L98 and matching code from read with this PR?
> 
> as we discussed offline, i think that two lines of code is useful that can save the sgl dup in daos obj layer in this PR. This pr can handle other case the passed in 0 iov although.

I made this comment before we discussed it off-line.  I agree that these two should stay but if you need to re-push you could update the comment.

**Review by mchaarawi — APPROVED** [review]

File: src/object/cli_obj.c (hunk 2 above)
Comment (soumagne) [reply src/object/cli_obj.c:4538]: @wangdi1 sorry I completely missed your question, yes it is allowed to pass a bulk handle that has NULL/0 lengths, this was requested long time ago.

Comment (wangdi1) [reply src/object/cli_obj.c:4538]: Thanks @soumagne , so @liuxuezhao  this failure might cause by some offset mis-calculation inside daos stack? or I miss sth?

**Pull request merged**